sklearn/cluster/tests/test_optics.py

```diff
@@ -0,0 +1,386 @@
+# Authors: Shane Grigsby <refuge@rocktalus.com>
+#          Amy X. Zhang <axz@mit.edu>
+# License: BSD 3 clause
+
+import numpy as np
+
+from sklearn.datasets.samples_generator import make_blobs
+from sklearn.cluster.optics_ import OPTICS
+from sklearn.cluster.optics_ import _TreeNode, _cluster_tree
+from sklearn.cluster.optics_ import _find_local_maxima
+from sklearn.utils.testing import assert_equal, assert_warns
+from sklearn.utils.testing import assert_array_almost_equal
+from sklearn.utils.testing import assert_array_equal
+from sklearn.utils.testing import assert_raises
+
+from sklearn.cluster.tests.common import generate_clustered_data
+
+
+def test_correct_number_of_clusters():
+    # in 'auto' mode
+
+    n_clusters = 3
+    X = generate_clustered_data(n_clusters=n_clusters)
+    # Parameters chosen specifically for this task.
+    # Compute OPTICS
+    clust = OPTICS(max_bound=5.0 * 6.0, min_samples=4, metric='euclidean')
+    clust.fit(X)
+    # number of clusters, ignoring noise if present
+    n_clusters_1 = len(set(clust.labels_)) - int(-1 in clust.labels_)
+    assert_equal(n_clusters_1, n_clusters)
+
+
+def test_minimum_number_of_sample_check():
+    # test that we check a minimum number of samples
+
+    # Compute OPTICS
+    X = [[1, 1]]
+    clust = OPTICS(max_bound=5.0 * 0.3, min_samples=10)
+
+    # Run the fit
+    assert_raises(ValueError, clust.fit, X)
```

Comment:
it would be better if you used `assert_raise_message` or `assert_raises_regexp` to be sure that the error is happening in the right place. though I suppose line coverage may be indicative.

Reply:
I can't seem to get `assert_raise_message` to work; perhaps because it's designed to work with errors and this is throwing a warning instead? When I try to implement in this test, I get an ```AssertionError: RuntimeWarning not raised by extract_dbscan``` , although clearly it is as the RuntimeWarning is printed in the same output... Do you have an example where `assert_raise_message` is used with a warning so I can check against that?

Reply:
I'm confused. Here you're checking for raising an error, not a warning. `assert_raise_message` works with errors.

`assert_warns` and `assert_warns_message` work with warnings.

Reply:
Was looking at the wrong line number-- this is fixed now for all error messages.